### DISCUSSION TITLE
[Feature request] Add copy mode

### DISCUSSION THREAD

**Discussion by mister-choo** [ghostty-org/ghostty#3488] [Feature Requests, Ideas]
Add a copy mode, like in wezterm, which allows to navigate the scrollback buffer with keyboard

**Comment by MironPascalCaseFan** [comment]
I would rather prefer to have Vim mode like in Alacrity https://github.com/alacritty/alacritty/blob/master/docs/features.md.
because with Wezterm copy mode it is impossible to achieve real Vim normal mode behavior.

**Reply by mister-choo** [reply]
Yea, obviously if there were even more functionality this would be nice.
For even more supercharged option there is https://github.com/mikesmithgh/kitty-scrollback.nvim , which allows for embedded neovim for scrollback buffer, which provides full vim functionality, without the need to implement it all from scratch

**Reply by MithicSpirit** [reply]
I've found that piping stuff into editors is substantially less powerful than having it built-in, as editors don't handle Very Long files very well.

**Comment by radiaku** [comment]
+1 , this mode good to have when need copy or moving around on terminal rather than use mouse. and yes vim mode alacritty its excellent Or iterms  copy mode. 

Its must have when on tmux doing copy on another buffer window and paste it on another window. Its become really akward if from mouse to keyboard and then keyboard again. ( atleast on my experience )

**Reply by 0x4a61636f62** [reply]
I think Alacritty's  Vim mode could be better. It would be amazing if Ghostty could somehow include a Vim mode to navigate the scrollback that has linenumbers.

**Comment by in43sh** [comment]
Can someone explain what exactly is copy mode?

**Reply by radiaku** [reply]
https://iterm2.com/documentation-copymode.html


Basically, you can select and copy text from the terminal output.

Really good when you are on tmux. Or you have spent a lot of time at the terminal. 

Other than that, I think you dont need it.

**Reply by mister-choo** [reply]
Personally I use it all the time to copy outputs of last commands, or selecting error, or other

**Comment by cynicaljoy** [comment]
This is pretty much the only feature keeping me from giving Ghostty a serious test drive. I have a lot of productivity gains from WezTerm [Quick Select Mode](https://wezfurlong.org/wezterm/quickselect.html).

**Reply by rmacklin** [reply]
WezTerm's "quick select mode" is covered by this older feature request (the title is based on the analogous feature in kitty):
- https://github.com/ghostty-org/ghostty/discussions/2394

**Comment by 00-kat** [comment]
Related: #3708.

**Comment by Stenstromen** [comment]
I would also like this feature, +1

**Reply by cynicaljoy** [reply]
you're getting 👎  responses because adding "+1" comments without any additional input as to what needs to be included in the feature is generally frowned upon. Just add your vote at the top.

**Comment by timursevimli** [comment]
My solution is to use tmux inside Ghostty, which provides vim-like copy-paste functionality through `prefix + [`

**Reply by cynicaljoy** [reply]
tmux `prefix + [` copy mode vs WezTerm copy mode is like iPod vs iPhone. Both can listen to music...but one can do a whole lot more.

**Reply by beef1218** [reply]
I do that too. But the problem for me is that I work with both my local macbook and a remote Linux machine. I rather launch Tmux after I sshed into my Linux machine. 
If I launch tmux on my local machine, I will have to launch another nested tmux after I sshed into my Linux.

**Reply by timursevimli** [reply]
> I do that too. But the problem for me is that I work with both my local macbook and a remote Linux machine. I rather launch Tmux after I sshed into my Linux machine. If I launch tmux on my local machine, I will have to launch another nested tmux after I sshed into my Linux.

By the way, to avoid such issues, I use CTRL + A as the prefix for tmux on my local MacBook, and it works well when the remote Linux machine uses the default prefix CTRL + B in tmux. This way, it becomes convenient to manage tmux opened on the remote machine through the local tmux session

**Comment by 00-kat** [comment]
It wasn't already noted here, but for the record the current workaround is the `write_screen_file:open` keybinding action.  If that doesn't work for you, see #3585.  You almost certainly want `write_screen_file` and not `write_scrollback_file`, see #3652 and #3496.

**Comment by V4G4X** [comment]
As per the workaround by @00-kat , I tried using the `write_screen_file:open` action.

But it keeps opening the scrollback in TextEdit (default MacOS text editor).  
Is there a simple way to set Neovim as the default text editor for .txt files?  
Maybe when I run that action it opens the buffer in another Ghostty window with Neovim running inside it.

Are you creating custom/dummy "automated" applications that do the above?  
I wanna get some ideas.

**Reply by V4G4X** [reply]
In this end, this is what I went with: https://github.com/qvacua/vimr  
You can also try: https://macvim.org/

Basically a macOS GUI for Neovim. (or vim if you use Macvim)  
Now I simply had to change the default opener for .txt to this app.  

if you don't wanna install a new app for this, 
and want the file path to be piped into Ghostty's Neovim with something as  
`ghostty -e neovim <filepath>`

You could check this: https://dialedinconfig.com/open-text-file-from-finder-in-neovim-or-vim/

**Reply by 00-kat** [reply]
>  Is there a simple way to set Neovim as the default text editor for .txt files?

Solutions from some other people on the [Ghostty Discord server] for macOS:
- [Someone's little swift script to do it in the most Apple-approved™ way][swift].
- [A command that someone has said works][cmd].
- [An application someone sent for doing this][app].

I don't use macOS so I've never tried nor can vouch for any of these solutions.  These are just ones I've seen floating around.

[Ghostty Discord server]: https://discord.gg/ghostty
[swift]: https://discord.com/channels/1005603569187160125/1324229986256818206/1324263042078347308
[cmd]: https://discord.com/channels/1005603569187160125/1323688219878887454/1323691381000241273
[app]: https://canary.discord.com/channels/1005603569187160125/1320882404717625374/1322438612205305943

A reproduction[^1] of the conversations linked above for those who cannot or do not wish to join the Ghostty Discord server:
<details>
<summary>Swift script</summary>

***Note**: there is quite a bit of text here, scroll to the bottom of this collapsed section for the actual Swift script that someone wrote up.*

### How can I change the the default config editor on macOS?

【2025-01-02 04:17 UTC】 **<ins>noe 💜</ins>**: When pressing Command+Comma ("Settings...") my configuration file is opened in TextEdit. This is not ideal. Is there a way to force it to open in a different editor when opened with Command+Comma? The file itself has no file extension so macOS itself cannot easily associate an application with it (to my understanding).
【2025-01-02 04:18 UTC】 **<ins>notjason</ins>**: in finder, option click on file, get info, then iirc you can set an app for just the config file
【2025-01-02 04:18 UTC】 **<ins>noe 💜</ins>**: I've done that, but the OS doesn't actually do anything for some reason 😦
【2025-01-02 04:18 UTC】 **<ins>noe 💜</ins>**: It still displays TextEdit as the default.
【2025-01-02 04:19 UTC】 **<ins>noe 💜</ins>**: Were you able to use that method?
【2025-01-02 04:19 UTC】 **<ins>notjason</ins>**: no i use nvim in the terminal to edit
【2025-01-02 04:19 UTC】 **<ins>noe 💜</ins>**: I see.
【2025-01-02 04:19 UTC】 **<ins>notjason</ins>**: sry :/
【2025-01-02 04:19 UTC】 **<ins>noe 💜</ins>**: No worries.
【2025-01-02 04:21 UTC】 **<ins>notjason</ins>**: hmmm that did work for me
![A screenshot demonstrating that the macOS setting worked for changing the default to Notes.app.](https://github.com/user-attachments/assets/429784d7-040b-4e99-b39d-ece661e4a249)
【2025-01-02 04:21 UTC】 **<ins>noe 💜</ins>**: Yeah, I change it, but then TextEdit is still listed as the default.
【2025-01-02 04:22 UTC】 **<ins>notjason</ins>**: 🫠
【2025-01-02 04:22 UTC】 **<ins>pluie</ins>**: possibly related to #2504?
【2025-01-02 04:22 UTC】 **<ins>Ghostty Bot</ins>**: (in reply to pluie's message) **Issue #2504:** `write_scrollback_file:open` to use `$EDITOR`
<https://github.com/ghostty-org/ghostty/issues/2504>
【2025-01-02 04:23 UTC】 **<ins>notjason</ins>**: i assume permissions are fine?
![A screenshot of the “Sharing & Permissions” of the file.  `jjt (Me)` can `Read & Write`, `staff` can `Read only`, while `everyone` can `Read only`.](https://github.com/user-attachments/assets/7c58aee2-9a1b-4258-8496-e9f0ae64f997)
【2025-01-02 04:23 UTC】 **<ins>noe 💜</ins>**: This is what I see.
![A screenshot of the “Open With” dialogue, with “Visual Studio Code” selected and “TextEdit (default)” next on the list.](https://github.com/user-attachments/assets/97eb7c93-e5d6-47fc-9a08-0a530aacc86b)
【2025-01-02 04:23 UTC】 **<ins>notjason</ins>**: oh yeah that's same here
【2025-01-02 04:24 UTC】 **<ins>notjason</ins>**:
![A similar screenshot, just with “Notes.app” selected instead of “Visual Studio Code”.](https://github.com/user-attachments/assets/4c86cf0e-66e5-4b96-8d0b-d6f545ae59cb)
【2025-01-02 04:24 UTC】 **<ins>notjason</ins>**: but double clicking on it opens in Notes.app
【2025-01-02 04:24 UTC】 **<ins>noe 💜</ins>**: Command+Comma in Ghostty, though?
【2025-01-02 04:24 UTC】 **<ins>notjason</ins>**: ohhhh let me see
【2025-01-02 04:24 UTC】 **<ins>notjason</ins>**: oh yeah still textedit
【2025-01-02 04:24 UTC】 **<ins>notjason</ins>**: mb sorry if i missed that context!
【2025-01-02 04:24 UTC】 **<ins>noe 💜</ins>**: Yeah
【2025-01-02 04:24 UTC】 **<ins>noe 💜</ins>**: No worries!
<sup> **<ins>notjason</ins>**: (reacting to the message above) 🙏</sup>
【2025-01-02 04:25 UTC】 **<ins>notjason</ins>**: welp at least it's not just you? 🙃
<sup> **<ins>noe 💜</ins>**: (reacting to the message above) 🙃</sup>
【2025-01-02 04:25 UTC】 **<ins>noe 💜</ins>**: (in replying to Ghostty Bot) not sure how it would be related, but $EDITOR feels like it might be a nice thing to use for the Command-Comma thing.
<sup> **<ins>notjason, nkonin, Hugo Gouveia, Kat</ins>**: (reacting to the message above) 💯</sup>
【2025-01-02 04:26 UTC】 **<ins>notjason</ins>**: i'd love that, right now i just `ctrl+r` and type `edit` until this comes up for me
![A screenshot of notjason's shell history demonstrating what he means.](https://github.com/user-attachments/assets/04f7457b-23db-448e-98d1-737e729cd233)

【2025-01-02 04:27 UTC】 **<ins>notjason</ins>**: i haven't gotten quite as far as adding a shell fn or alias, but i'm close!
【2025-01-02 04:28 UTC】 **<ins>notjason</ins>**: i do miss this from kitty, `cmd+,` opens a new window with `$EDITOR` (nvim for me)
【2025-01-02 04:38 UTC】 **<ins>noe 💜</ins>**: Found the open handler: https://github.com/ghostty-org/ghostty/blob/v1.0.1/src/config/edit.zig#L19

Now to figure out why it's behaving this way.
【2025-01-02 04:43 UTC】 **<ins>noe 💜</ins>**: Found it.
https://github.com/ghostty-org/ghostty/blob/v1.0.1/src/os/open.zig#L35

It runs `open -t ${configPath}`. 

The man page (https://www.unix.com/man-page/osx/1/open/, unofficial source) case the `-t` option "Causes the file to be opened with the default text editor, as determined via LaunchServices".
【2025-01-02 04:44 UTC】 **<ins>noe 💜</ins>**: So it seems I just need to register the default text editor. But it seems macOS doesn't provide a nice way to do that? I'll have to look into it.
【2025-01-02 04:46 UTC】 **<ins>Kat</ins>**: one sec lemme find a link to some instructions someone else wrote up which had worked for them
【2025-01-02 04:47 UTC】 **<ins>noe 💜</ins>**: Thanks!
【2025-01-02 04:48 UTC】 **<ins>Greg</ins>**: [*A message link to the command mentioned in the bullet points near the top of this comment.*][cmd]
【2025-01-02 04:49 UTC】 **<ins>noe 💜</ins>**: Nice. Given what I see when decompiling the `open` command, only the first command may be needed?
【2025-01-02 04:50 UTC】 **<ins>Greg</ins>**: correct
【2025-01-02 04:50 UTC】 **<ins>Greg</ins>**: If you want to use VS code that is
【2025-01-02 04:50 UTC】 **<ins>Greg</ins>**: Only reason I didn't provide the command was its dependent on which program you want to use
【2025-01-02 04:50 UTC】 **<ins>noe 💜</ins>**: What I meant was the `LSHandlerContentType=public.data` command is superfluous.
【2025-01-02 04:51 UTC】 **<ins>noe 💜</ins>**: the `open` command doesn't check for handlers for that content type when you use the `-t` command.
【2025-01-02 04:51 UTC】 **<ins>Kat</ins>**: (in reply to her previous message) this is the one i was looking for [*A message link to the application mentioned in the bullet points near the top of this comment.*][app]
【2025-01-02 04:52 UTC】 **<ins>noe 💜</ins>**: The file itself may be `public.data` but that's not taken into account when running `open -t`.
【2025-01-02 04:52 UTC】 **<ins>Kat</ins>**: (in reply to Greg's message link) (i'll also note this one down ig)
【2025-01-02 04:53 UTC】 **<ins>noe 💜</ins>**:  I'm really just trying to find a no-nonsense method and I think the first one might be best? I don't care about a GUI such as SwiftDefaultApps has.
【2025-01-02 04:53 UTC】 **<ins>noe 💜</ins>**: A direct first-party way to tell the OS "Hey! This is my text editor!"
【2025-01-02 04:54 UTC】 **<ins>noe 💜</ins>**: Using `open` does kind of make using `$EDITOR` impossible, though (as it needs to be an App, not just a binary). But it makes sense, since that is the OS-level app open tool.
【2025-01-02 04:55 UTC】 **<ins>Kat</ins>**: i assume that no shell is packaged as an `.app`?
【2025-01-02 04:55 UTC】 **<ins>Greg</ins>**: Yeah my linked method is the most direct way as I generally prefer to be direct over aliases
【2025-01-02 04:55 UTC】 **<ins>Greg</ins>**: or abstractions
【2025-01-02 04:56 UTC】 **<ins>Kat</ins>**: because something funky like `open -t SomeShellOrSomething.app -c $EDITOR` might work
【2025-01-02 04:56 UTC】 **<ins>noe 💜</ins>**: Well, from a technical perspective it's just any binary that can handle whatever `open` throws at it.
【2025-01-02 04:57 UTC】 **<ins>noe 💜</ins>**: (in reply to Kat) This would be two layers of command indirection though, lol.
【2025-01-02 04:57 UTC】 **<ins>Kat</ins>**: tru
【2025-01-02 04:57 UTC】 **<ins>noe 💜</ins>**: Honestly, even calling `open` as `std.process` raises my hairs a tad.
【2025-01-02 04:58 UTC】 **<ins>noe 💜</ins>**: But that's just me.
【2025-01-02 05:07 UTC】 **<ins>noe 💜</ins>**: Wait, it's passing it in as an array of args. Ok, that should be fine.
【2025-01-02 06:28 UTC】 **<ins>noe 💜</ins>**: Here's a quick no-nonsense Swift script I wrote. It uses the Apple-reccomended API's. The other options are cool, but I think they might be a bit side-effect-y.
```swift
import AppKit

guard CommandLine.argc > 1,
    case let action = CommandLine.arguments[1],
    (action == "get" && CommandLine.argc == 2) || (action == "set" && CommandLine.argc == 3)
else {
    print("Usage: \(CommandLine.arguments[0]) get | set <path>")
    exit(1)
}

switch action {
case "get":
    guard let url = NSWorkspace.shared.urlForApplication(toOpen: .plainText) else {
        print("No default application for plain text files, or was not found.")
        exit(0)
    }
    let bundle = Bundle(url: url)
    print(
        "The default application for plain text files is \(bundle?.bundleIdentifier ?? url.path).")
    exit(0)
case "set":
    let path = CommandLine.arguments[2]
    NSWorkspace.shared.setDefaultApplication(at: URL(filePath: path), toOpen: .plainText) {
        err in
        if let actualErr = err {
            print(actualErr)
            exit(1)
        }
        let bundle = Bundle(path: path)
        print(
            "The default application for plain text files is now \(bundle?.bundleIdentifier ?? path)."
        )
    }
default:
    print("Unknown action: \(action)")
    exit(1)
}
```
【2025-01-02 06:29 UTC】 **<ins>Greg</ins>**: Well thats one way to do it lol.
【2025-01-02 06:30 UTC】 **<ins>noe 💜</ins>**: I prefer it. Writing entries to an undocumented preference dictionary (even if it works) makes me feel a bit weird.
<sup> **<ins>Kat</ins>**: (reacting to the message above) ![Emoji, real](https://github.com/user-attachments/assets/a9f63959-c6fa-4c5e-93b3-8800ed7e68af)</sup>
【2025-01-02 06:31 UTC】 **<ins>Greg</ins>**: Understandable
【2025-01-02 06:35 UTC】 **<ins>Kat</ins>**: o kewl
【2025-01-02 06:37 UTC】 **<ins>noe 💜</ins>**: The one thing is this *is* somewhat-based off of me reverse-engineering the `open` command. At the time of writing, it uses the `plainText` type to get the application to open.
【2025-01-02 06:37 UTC】 **<ins>noe 💜</ins>**: I don't doubt that will ever change. But the fact that it does that *specificially* is technically undocumented. There is another file type called `text`, but it doesn't use that one.
<sup> **<ins>Kat</ins>**: (reacting to the message above) 👀</sup>
【2025-01-02 08:45 UTC】 **<ins>noe 💜</ins>**: Additional clarification, the `open` API also uses a deprecated function. The "get" function in my script above is the documented replacement. The "set" function is an educated guess at the corollary function.
<sup> **<ins>Kat</ins>**: (reacting to the message above) ![Emoji, WHAT](https://github.com/user-attachments/assets/84164722-4c1f-4e5f-8eb9-766fa239607b)</sup>
【2025-01-02 08:45 UTC】 **<ins>noe 💜</ins>**: All in all though, I think it's the best modern solution we have, and it works, so I'm happy with it.
【2025-01-02 08:57 UTC】 **<ins>noe 💜</ins>**: Yes, there's a lot of context and I'm probably overly explaining. My autistic ass tends to do that, lol.
<sup> **<ins>Kat</ins>**: (reacting to the message above) ![Emoji, real](https://github.com/user-attachments/assets/a9f63959-c6fa-4c5e-93b3-8800ed7e68af)</sup>
【2025-01-02 08:58 UTC】 **<ins>noe 💜</ins>**: In short, yes to the best of my knowledge my script works and probably will work long into the foreseeable future. In long: see above.

</details>

<details>
<summary>Command</summary>

### Change program super+, opens on MacOS

【2024-12-31 16:24 UTC】 **<ins>joeriddles</ins>**: Is it possible to change the program that `super+,` opens on MacOS? Currently it opens the config in TextEdit but I would prefer to open `nano` or VS Code.
【2024-12-31 16:30 UTC】 **<ins>Greg</ins>**: You will have to change your systems default text editor
【2024-12-31 16:35 UTC】 **<ins>joeriddles</ins>**: Thanks! Tried these commands, so far no dice:
```console
$  defaults write com.apple.LaunchServices/com.apple.launchservices.secure \
    LSHandlers -array-add \
    '{LSHandlerContentType=public.plain-text;LSHandlerRoleAll=com.microsoft.VSCode;}'

$ defaults write com.apple.LaunchServices/com.apple.launchservices.secure \
    LSHandlers -array-add \
    '{LSHandlerContentType=public.data;LSHandlerRoleAll=com.microsoft.VSCode;}'
```
【2024-12-31 16:36 UTC】 **<ins>joeriddles</ins>**: Appears I have the right content type:
```console
$ mdls -name kMDItemContentType .config/ghostty/config
kMDItemContentType = "public.data"
```
【2024-12-31 16:45 UTC】 **<ins>joeriddles</ins>**: Restarted my computer and now it works!
【2024-12-31 16:51 UTC】 **<ins>Greg</ins>**: Please go ahead and mark this thread as solved!

</details>

<details>
<summary>Application</summary>

【2024-12-28 05:38 UTC】**<ins>DarkHorse</ins>**: On macOS, to change default app that is used to edit the ghostty config file (Cmd+,):
The desired editor must be a real app, cannot be a terminal command.  So, for vim, install https://macvim.org/
Download and install this Preference Pane utility  https://github.com/Lord-Kamina/SwiftDefaultApps (unzip, double click .prefpane to install)
Go into System Settings, scroll to the bottom, select the newly added SwiftDefaultApps pane
Select Uniform Type Identifiers tab. The ghostty config file has no extension, so it’s data type should be public.data. Select public.data from the list. 
On the right side of the pane specify the desired editor app. For vim, select MacVim.app
Also select public.plain-text from the list and specify the desired editor app.
That’s it.  When you hit Cmd+, in ghostty, or select Ghostty->Settings... from the app menu, your editor should now launch.

</details>


[^1]: I am  not responsible for any of the reproduced messages or their content, and any copyright associated with them belongs to the author of the message.  See the message links I sent near the top to view the original messages.

-----

And for Linux (I've tested all of these):
- Run `xdg-mime default yourfavouriteeditor.desktop text/plain`.  You can find a list of desktop files on your system with `find /usr/share/applications ~/.local/share/applications -name '*.desktop'`.
- Depending on your file manager, you may be able to right-click on the Ghostty configuration file and choose your favourite editor from the list.
  <details>
  <summary>Click or tap for an example screenshot in Thunar.</summary>

  ![A screenshot of the Thunar right-click menu when right-clicking on the Ghostty configuration file.  ‘Open With → Set Default Application…’ is selected.](https://github.com/user-attachments/assets/2a9dc7cc-81b1-4dda-8beb-16df2104ad93)
  </details>
- Edit `~/.config/mimeapps.list` manually.  The file format is an atrocious monstrosity basking in INI's glow, so I can't give a consise description of how this is done, sorry.  Prefer using the above two solutions, they edit this file themselves too.

Please note that this won't work with TUI editors without editing the desktop file, see #3585 (or more specifically, https://github.com/ghostty-org/ghostty/issues/2435#issuecomment-2564058758).

**Comment by purajit** [comment]
Found this because of a comment on my feature request at https://github.com/ghostty-org/ghostty/discussions/7672.

This discussion doesn't exactly cover this, but posting here since it is adjacent. Would love if we had something like tmux-fingers (see that discussion for an explanation)! It wouldn't be a subfunction of copy mode, since that would kind of defeat the purpose, but there would likely be good overlap in the implementation.

**Reply by rmacklin** [reply]
I haven't used tmux-fingers, but what you described sounds more like this feature request: https://github.com/ghostty-org/ghostty/discussions/2394 (inspired by [kitty's "hints" feature](https://sw.kovidgoyal.net/kitty/kittens/hints/))

**Reply by purajit** [reply]
That does indeed seem to be the right place for this

**Comment by saifulapm** [comment]
Any updates on this ?

**Comment by jzbyers** [comment]


**Reply by MithicSpirit** [reply]
I don't think that there needs to be any special mechanisms for pasting the copied selection (other than the normal interaction with the keyboard).
Also, for further clarification, I think that the navigation/selection should be similar to vim's normal/visual modes, in that one can move use keybinds to select start and end positions for copying (though it may be better for default keybindings to use the arrow keys rather than hjkl). It would be nice to also support visual line and visual block mode, but I would be happy even without those at this point.

**Reply by rhodes-b** [reply]
Are you using AI to generate these prompts because there are conflicting statements in your proposal
> Key aspects:
Copy operation writes directly to system clipboard

then right after 
> super+c to copy selection to clipboard (default behavior in Ghostty)

Also search currently does not work in ghostty in any mode so it definetly wont work in copy mode. Search is more than a big enough feature itself they should be entirely separate. I also dont see why the normal search keybind (ctrl + f) also wouldn't work even if other copy modes don't support that

> Esc or q to cancel/exit copy mode

This will have familiarity but it doesn't match any other toggle action in ghostty where the same keybind is used to enter / exit a state like `toggle_tab_overview`, `toggle_fullscreen`, `toggle_terminal_inspector` there are more I just named a few. Note that esc on the tab overview also exits it but that is handled by the adwaita widget not ghostty im pretty sure. The point im making is it should be a toggle keybind so I can exit with the keybind that toggled it in the first place to match existing ghostty behavior.

> shift+←/→ to expand selection horizontally
shift+↑/↓ to expand selection vertically

Your keybinds for starting selection also dont match tmux behavior. Although im not sure how any other copy mode works. In tmux you press a bind to start the selection then normal cursor movements continue that selection I dont need to keep pressing a extend selection key.

Just as a warning / heads up if you do start on this. If you go guns blazing out of the gates and implement everything suggested above in one PR it will be rejected without any review (especially since this is not a fully accepted proposal yet although the amount of people who want it is pretty large so there would be no reason to reject it outright). Think of the perspective of the maintainers small changes are better in every way as its easier to review and verify that the behavior is good, it will build your trust of contributions. Getting a basic state where you enter a mode where you can navigate the scrollback with keys is enough for the first PR then follow ups can add selection handling, then alternate keybinds etc, then sometime in the future when search is added hooking up search (if this landed first)

Just as a note since I do believe your prompts have been AI generated. Ghostty has a clear policy on disclosing AI usage in contributions even just consulting it to learn about the codebase. https://github.com/ghostty-org/ghostty/blob/main/HACKING.md#ai-and-agents
Ghostty is not against AI contributions Mitchell is a big fan of using AI.

**Reply by MithicSpirit** [reply]
> Your keybinds for starting selection also dont match tmux behavior. Although im not sure how any other copy mode works. In tmux you press a bind to start the selection then normal cursor movements continue that selection I dont need to keep pressing a extend selection key.

Tmux behavior would correspond to vim mode I think.

**Reply by jzbyers** [reply]
I have a WIP branch of this working but the UX doesn't feel very "ghostty" or platform-native to me so I'm deleting my comments above, which in hindsight feel too prescriptive and too heavily influenced by the copy mode features I'm used to using in alacritty + tmux.

**Comment by rekmarks** [comment]
A copy mode modeled on that of tmux (i.e. vim's visual mode) is the last thing that's preventing me from uninstalling tmux and just using ghostty natively for everything. Running tmux in ghostty is a bit like putting the engine of a Ford Model T inside a Mustang, so getting this feature would be amazing.
